dlib/threads/thread_pool_extension_abstract.h

```diff
@@ -234,11 +298,17 @@ namespace dlib
 
     public:
         explicit thread_pool (
-            unsigned long num_threads
+            unsigned long num_threads,
+            thread_pool_thread_manager* manager = nullptr
```

Comment:
This all looks sensible.  But why not use a std::shared_ptr or a std::unique_ptr since it sounds like a manager shouldn't be shared between pools.

Reply:
It's fine to share managers between pools if the manager is built to handle that. In my code running in the browser, I indeed have exactly one manager that is passed to every thread pool. But that is a controlled situation.

The default implementation for when no custom manager is supplied is however naive. It just forever adds to a `std::vector<std::thread>` and never cleans up until destruction. When using one instantiation per pool that is not a problem (exactly num_threads are created once at thread pool creation and joined once at thread pool shutdown, after which the object is destroyed, the vector is never larger than it needs to be), but when sharing it'll grow to the size of num_threads_created_ever. Of course that could be corrected, but it's not necessary in the current implemenation, and I wanted to keep it as simple as possible as C++ is a trip outside my regular programming.

Furthermore I wasn't sure if a thread in a pool can in turn create its own pool, in which case a reused manager would have to be thread safe (with the associated cost), which it doesn't have to be if every pool creates its own instance, as the method are then always called from the same thread.

If you still think shared_ptr or unique_ptr is the way to go, I guess I'll have to read up on those :)

Reply:
Yeah that's fine.  I just mean, use a `std::shared_ptr` so that ownership of the pointer is clear.  E.g. then you don't have to call `delete`.  You should never have to manually call delete in C++ :D